packages/core/database/src/index.ts

```diff
@@ -97,8 +100,30 @@ class Database {
   }
 
   async init({ models }: { models: Model[] }) {
+    /**
+     * TODO v6: implement a cleaner solution that will involve breaking changes to db constructor
+     *
+     * if the user passed in a connection function, we need to resolve it first
+     * and recreate the connection, otherwise accessing this.connection.anyProperty will not work
+     *
+     * Note that even if it was already created in the constructor, the connection method will
+     * not be called at this point by Strapi or Knex, so it shouldn't have any effects to do this
+     *  */
+    if (typeof this.config.connection.connection === 'function') {
+      debug('User config.connection is a function; resolving and recreating knex connection');
+      this.config.connection.connection = await this.config.connection.connection();
```

Comment:
Final version we would clone or deconstruct the user config instead of rewriting it